// ngx-showdown/esm5/showdown.pipe.js
import * as tslib_1 from "tslib";
import { Pipe } from '@angular/core';
import { ShowdownConverter } from './showdown-converter.provider';
/**
 * A angular pipe to transform `Markdown` to `HTML`.
 *
 * ### Example
 *
 * Setup as standalone
 * ```typescript
 * import { NgModule } from '@angular/core';
 * import { ShowdownPipe } from 'ngx-showdown';
 *
 * @NgModule({
 *   declarations: [ ShowdownPipe ];
 * })
 * export class AppModule {}
 * ```
 *
 * Transform markdown value to html.
 * ```html
 * <input type="text" placeholder="Name" [(ngModel)]="name"/>
 * <div [innerHTML]="'**Hello '+(name || 'nobody')+'!**' | showdown">
 * ```
 *
 * Transform markdown value to html with `options`.
 * ```typescript
 * import * as Showdown from 'showdown';
 *
 * @Component({
 *   selector: 'some',
 *   template: `<div innerHTML="{{ text | showdown: options }}"></div>`
 * })
 * export class SomeComponent {
 *   text: string = "__Some Underline__";
 *   options: Showdown.ShowdownOptions = { underline: true };
 *   // ...
 * }
 * ```
 */
var ShowdownPipe = /** @class */ (function (_super) {
    tslib_1.__extends(ShowdownPipe, _super);
    function ShowdownPipe() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Transform markdown value to html.
     *
     * @param value - The markdown value to transform to html.
     * @param options - A `Showdown` converter options.
     * @returns Returns the transform result of `value` to html.
     */
    ShowdownPipe.prototype.transform = function (value, options) {
        this.setOptions(options);
        return this.makeHtml(value);
    };
    ShowdownPipe = tslib_1.__decorate([
        Pipe({
            name: 'showdown',
            pure: false
        })
    ], ShowdownPipe);
    return ShowdownPipe;
}(ShowdownConverter));
export { ShowdownPipe };
